File: Data_Structure/Cprogramming/ch8_Reculsive_calling/subset_sum.c
#include <stdio.h>
#include <stdlib.h>

int *data;
int *flag;
int sum = 0;
int count = 0;

void subset(int n, int depth, int s)
{
	if (n == depth)
	{
		int i;
		for (i = 0; i<n; i++)
		{
			if (flag[i] == 1)
			{
				sum += data[i];
			}
		}
		if (sum == s)
			count++;
		sum = 0;
		return;
	}
	flag[depth] = 1;
	subset(n, depth + 1, s);
	flag[depth] = 0;
	subset(n, depth + 1, s);

}

int main()
{
	int integer;
	int num;
	int i;

	printf("정수의 개수: ");
	scanf("%d", &num);

	printf("원소들의 합이 될 정수 :");
	scanf("%d", &integer);

	data = (int*)malloc(sizeof(int)*num);
	flag = (int*)calloc(num, sizeof(int));

	for (i = 0; i < num; i++)
	{
		printf("원소 : ");
		scanf("%d", &(*(data+i)));
	}

	subset(num, 0, integer);

	if (integer == 0)
		count--;

	printf("count = %d\n", count);

	free(data);
	free(flag);

}

File: Data_Structure/Problem/ch8_Reculsive_calling/subset_sum.py
def subset(data, position, n, depth, s):
    sum = 0
    if depth == 0:
        global count
        count = 0
    if n == depth :         # 재귀에서 나오기 위한 조건문
        for i in range(n):
            if position[i] == 1 :
                sum += data[i]
        if sum == s:
            count += 1
        return
    position[depth] = 1
    subset(data, position, n, depth + 1, s)
    position[depth] = 0
    subset(data, position, n, depth + 1, s)

data = []
position = []

num = int(input('정수의 개수 : '))
integer = int(input('원소들의 합이 될 정수 :'))

for i in range(num):
    x = int(input('원소 : '))
    data.append(x)

for i in range(num):
    position.append(0)

subset(data, position, num, 0, integer)

if (integer == 0):
    count -= 1

print('{}이/가 되는 부분집합의 개수는 {} 이다.'.format(integer,count))